File: js/algorithms.js
// ============================================================
// STRUKTUR DATA & ALGORITMA — Client-side JavaScript
// Graf, BFS, Dijkstra, Stack (Linked List), BST, Merge Sort
// ============================================================

// ---- Build Graf ----
const halteMap = {};
const adj = {};

function initGraf() {
  for (const [hid, nama, zona, fas] of DATA_HALTE) {
    halteMap[hid] = { id: hid, nama, zona, fasilitas: fas };
    adj[hid] = [];
  }
  for (const [a, b, d] of DATA_EDGE) {
    adj[a].push([b, d]);
    adj[b].push([a, d]);
  }
}

// ---- BFS: Rute Halte Tersedikit ----
function bfs(asal, tujuan) {
  if (!(asal in halteMap) || !(tujuan in halteMap))
    return { path: null, jarak: null, ms: null };

  const t0 = performance.now();
  const visited = new Set([asal]);
  const queue = [[asal, [asal]]];
  let head = 0;

  while (head < queue.length) {
    const [curr, path] = queue[head++];
    if (curr === tujuan) {
      const ms = performance.now() - t0;
      return { path, jarak: hitungJarak(path), ms: +ms.toFixed(4) };
    }
    for (const [nid] of adj[curr]) {
      if (!visited.has(nid)) {
        visited.add(nid);
        queue.push([nid, [...path, nid]]);
      }
    }
  }
  return { path: null, jarak: null, ms: null };
}

// ---- Dijkstra: Rute Jarak Terpendek ----
function dijkstra(asal, tujuan) {
  if (!(asal in halteMap) || !(tujuan in halteMap))
    return { path: null, jarak: null, ms: null };

  const t0 = performance.now();
  const dist = {};
  const prev = {};
  for (const hid in halteMap) {
    dist[hid] = Infinity;
    prev[hid] = null;
  }
  dist[asal] = 0;

  // Simple priority queue using sorted array
  const heap = [[0, asal]];
  while (heap.length) {
    heap.sort((a, b) => a[0] - b[0]);
    const [d, curr] = heap.shift();
    if (d > dist[curr]) continue;
    if (curr === tujuan) break;
    for (const [nid, w] of adj[curr]) {
      const alt = dist[curr] + w;
      if (alt < dist[nid]) {
        dist[nid] = alt;
        prev[nid] = curr;
        heap.push([alt, nid]);
      }
    }
  }

  const ms = performance.now() - t0;
  if (dist[tujuan] === Infinity)
    return { path: null, jarak: null, ms: null };

  const path = [];
  let node = tujuan;
  while (node !== null) {
    path.unshift(node);
    node = prev[node];
  }

  return { path, jarak: +dist[tujuan].toFixed(2), ms: +ms.toFixed(4) };
}

function hitungJarak(path) {
  let total = 0;
  for (let i = 0; i < path.length - 1; i++) {
    const curr = path[i];
    const nxt = path[i + 1];
    for (const [nid, d] of adj[curr]) {
      if (nid === nxt) { total += d; break; }
    }
  }
  return +total.toFixed(2);
}

// ---- Stack berbasis Linked List ----
class StackNode {
  constructor(val) {
    this.val = val;
    this.next = null;
  }
}

class LinkedStack {
  constructor(maxSize = 20) {
    this.head = null;
    this.size = 0;
    this.maxSize = maxSize;
  }

  push(val) {
    const node = new StackNode(val);
    node.next = this.head;
    this.head = node;
    this.size++;
    if (this.size > this.maxSize) {
      let curr = this.head;
      while (curr.next && curr.next.next) curr = curr.next;
      if (curr.next) curr.next = null;
      this.size--;
    }
  }

  pop() {
    if (!this.head) return null;
    const val = this.head.val;
    this.head = this.head.next;
    this.size--;
    return val;
  }

  toList() {
    const res = [];
    let curr = this.head;
    while (curr) { res.push(curr.val); curr = curr.next; }
    return res;
  }

  clear() {
    this.head = null;
    this.size = 0;
  }
}

const historyStack = new LinkedStack(20);

// ---- Merge Sort ----
function mergeSort(arr, keyFunc = null) {
  if (!keyFunc) keyFunc = x => x;
  if (arr.length <= 1) return arr;
  const mid = Math.floor(arr.length / 2);
  const L = mergeSort(arr.slice(0, mid), keyFunc);
  const R = mergeSort(arr.slice(mid), keyFunc);

  const res = [];
  let i = 0, j = 0;
  while (i < L.length && j < R.length) {
    if (keyFunc(L[i]) <= keyFunc(R[j])) res.push(L[i++]);
    else res.push(R[j++]);
  }
  while (i < L.length) res.push(L[i++]);
  while (j < R.length) res.push(R[j++]);
  return res;
}

// ---- BST (Binary Search Tree) ----
class BSTNode {
  constructor(key, val) {
    this.key = key.toLowerCase();
    this.val = val;
    this.left = null;
    this.right = null;
  }
}

class BST {
  constructor() { this.root = null; }

  insert(key, val) {
    this.root = this._insert(this.root, key, val);
  }
  _insert(root, key, val) {
    if (!root) return new BSTNode(key, val);
    if (key.toLowerCase() < root.key) root.left = this._insert(root.left, key, val);
    else root.right = this._insert(root.right, key, val);
    return root;
  }

  search(key) {
    return this._search(this.root, key.toLowerCase());
  }
  _search(root, key) {
    if (!root) return null;
    if (root.key === key) return root.val;
    return key < root.key ? this._search(root.left, key) : this._search(root.right, key);
  }

  prefixSearch(prefix) {
    const res = [];
    this._prefixSearch(this.root, prefix.toLowerCase(), res);
    return res;
  }
  _prefixSearch(root, prefix, res) {
    if (!root) return;
    this._prefixSearch(root.left, prefix, res);
    if (root.key.startsWith(prefix)) res.push(root.val);
    this._prefixSearch(root.right, prefix, res);
  }
}

const bstInstance = new BST();
function initBST() {
  for (const [hid, nama, zona, fas] of DATA_HALTE) {
    bstInstance.insert(nama, { id: hid, nama, zona, fasilitas: fas });
  }
}

File: algorithms.py
# ============================================================
# STRUKTUR DATA & ALGORITMA
# Graf, BFS, Dijkstra, Stack, BST, Merge Sort
# ============================================================

import time
import heapq
from collections import deque
from data import DATA_HALTE, DATA_EDGE

# ---- Build Graf ----
halteMap = {}
adj = {}

def init_graf():
    global halteMap, adj
    halteMap.clear()
    adj.clear()
    for row in DATA_HALTE:
        hid, nama, zona, fas = row
        halteMap[hid] = {
            "id": hid,
            "nama": nama,
            "zona": zona,
            "fasilitas": fas
        }
        adj[hid] = []
    
    for a, b, d in DATA_EDGE:
        adj[a].append((b, d))
        adj[b].append((a, d))

# Run initialization once on import
init_graf()

# ---- BFS: Rute Halte Tersedikit ----
def bfs(asal, tujuan):
    asal, tujuan = int(asal), int(tujuan)
    if asal not in halteMap or tujuan not in halteMap:
        return {"path": None, "jarak": None, "ms": None}
    
    t0 = time.perf_counter()
    visited = {asal}
    queue = deque([(asal, [asal])])
    
    while queue:
        curr, path = queue.popleft()
        if curr == tujuan:
            ms = (time.perf_counter() - t0) * 1000
            return {
                "path": path,
                "jarak": hitung_jarak(path),
                "ms": round(ms, 4)
            }
        for nid, _ in adj[curr]:
            if nid not in visited:
                visited.add(nid)
                queue.append((nid, path + [nid]))
                
    return {"path": None, "jarak": None, "ms": None}

# ---- Dijkstra: Rute Jarak Terpendek ----
def dijkstra(asal, tujuan):
    asal, tujuan = int(asal), int(tujuan)
    if asal not in halteMap or tujuan not in halteMap:
        return {"path": None, "jarak": None, "ms": None}
    
    t0 = time.perf_counter()
    dist = {hid: float('inf') for hid in halteMap}
    prev = {hid: None for hid in halteMap}
    dist[asal] = 0.0
    
    heap = [(0.0, asal)]
    while heap:
        d, curr = heapq.heappop(heap)
        if d > dist[curr]:
            continue
        if curr == tujuan:
            break
        for nid, w in adj[curr]:
            alt = dist[curr] + w
            if alt < dist[nid]:
                dist[nid] = alt
                prev[nid] = curr
                heapq.heappush(heap, (alt, nid))
                
    ms = (time.perf_counter() - t0) * 1000
    if dist[tujuan] == float('inf'):
        return {"path": None, "jarak": None, "ms": None}
        
    path = []
    node = tujuan
    while node is not None:
        path.insert(0, node)
        node = prev[node]
        
    return {
        "path": path,
        "jarak": round(dist[tujuan], 2),
        "ms": round(ms, 4)
    }

def hitung_jarak(path):
    total = 0.0
    for i in range(len(path) - 1):
        curr = path[i]
        nxt = path[i+1]
        for nid, d in adj[curr]:
            if nid == nxt:
                total += d
                break
    return round(total, 2)

# ---- Stack berbasis Linked List ----
class StackNode:
    def __init__(self, val):
        self.val = val
        self.next = None

class LinkedStack:
    def __init__(self, max_size=20):
        self.head = None
        self.size = 0
        self.max_size = max_size

    def push(self, val):
        node = StackNode(val)
        node.next = self.head
        self.head = node
        self.size += 1
        if self.size > self.max_size:
            # Remove the oldest (last) node
            curr = self.head
            while curr.next and curr.next.next:
                curr = curr.next
            if curr.next:
                curr.next = None
            self.size -= 1

    def pop(self):
        if self.head is None:
            return None
        val = self.head.val
        self.head = self.head.next
        self.size -= 1
        return val

    def to_list(self):
        res = []
        curr = self.head
        while curr:
            res.append(curr.val)
            curr = curr.next
        return res

    def clear(self):
        self.head = None
        self.size = 0

# Shared stack instance for the API search history
history_stack = LinkedStack(max_size=20)

# ---- Merge Sort ----
def merge_sort(arr, key_func=None):
    if key_func is None:
        key_func = lambda x: x
    if len(arr) <= 1:
        return arr
    mid = len(arr) // 2
    L = merge_sort(arr[:mid], key_func)
    R = merge_sort(arr[mid:], key_func)
    
    res = []
    i = j = 0
    while i < len(L) and j < len(R):
        if key_func(L[i]) <= key_func(R[j]):
            res.append(L[i])
            i += 1
        else:
            res.append(R[j])
            j += 1
    res.extend(L[i:])
    res.extend(R[j:])
    return res

# ---- BST (Binary Search Tree) ----
class BSTNode:
    def __init__(self, key, val):
        self.key = key.lower()
        self.val = val
        self.left = None
        self.right = None

class BST:
    def __init__(self):
        self.root = None

    def insert(self, key, val):
        self.root = self._insert(self.root, key, val)

    def _insert(self, root, key, val):
        if root is None:
            return BSTNode(key, val)
        if key.lower() < root.key:
            root.left = self._insert(root.left, key, val)
        else:
            root.right = self._insert(root.right, key, val)
        return root

    def search(self, key):
        return self._search(self.root, key.lower())

    def _search(self, root, key):
        if root is None:
            return None
        if root.key == key:
            return root.val
        if key < root.key:
            return self._search(root.left, key)
        return self._search(root.right, key)

    def prefix_search(self, prefix):
        res = []
        self._prefix_search(self.root, prefix.lower(), res)
        return res

    def _prefix_search(self, root, prefix, res):
        if root is None:
            return
        self._prefix_search(root.left, prefix, res)
        if root.key.startswith(prefix):
            res.append(root.val)
        self._prefix_search(root.right, prefix, res)

# Initialize BST with Biskita halts
bst_instance = BST()
for hid, nama, zona, fas in DATA_HALTE:
    halt_info = {"id": hid, "nama": nama, "zona": zona, "fasilitas": fas}
    bst_instance.insert(nama, halt_info)

def bst_search(nama):
    return bst_instance.search(nama)

def bst_prefix(prefix):
    return bst_instance.prefix_search(prefix)
